### ISSUE TITLE
[feature request] Easy way to upgrade/patch

### ISSUE THREAD

**Issue by mmahmoudian** [ismayc/thesisdown#90]
## General description

As we all know, writing a thesis takes time, and throughout the course of writing the thesis, this package will eventually get better and better and some bug and issues get fixed. There must be a painless way up upgrade a current code-base to the newer version.

## The issue in action
Just after #86 gets resolved, all the following files are now changed in multiple places:

```
.
└── index
    ├── _bookdown.yml
    ├── index.Rmd
    ├── reedthesis.cls
    └── template.tex
```

as both `_bookdown.yml` and `index.Rmd are the files that user should manually change to customize for their thesis, there are quite a few lines that the user should manually merge every time this package is getting updated.

## The feature I'm looking for

It would be super nice to have a function in the package that can "patch" the current thesis' files according to the new version of ThesisDown. The patch function can always have some sort of diff patch to exactly know what should be patched (from what to what) and to make sure we are not overwriting user settings, the function always checks "from what" and if it matched it will patch it. of course if the patch breaks something related to compiling/knitting, user can always checkout previous commit. 

 The way user interacts with it:
1. User updates the package to latest stable version:
    - `remotes::install_github("ismayc/thesisdown")`
2. User run the patch function:
    - `thesisdown::patch(path = "path/to/the/index/folder", verbose = FALSE)`
        - I just thought having a togglable verbosity would be nice so that user knows which files have changed.
        - the path can be by default the working directory.
3. user will be prompted to `git add . ; git commit` just to be safe
4. if there is any conflicts, user will be informed

**Comment by ismayc** [comment]
This sounds amazing, but also incredibly hard to implement. Among other things, `index.Rmd` is customized and tweaked a lot by people building their own theses. {thesisdown} has changed very little over the last four years really. It was designed to not have frequent updates in much the same way that {bookdown} was.

Just to be clear, I’m not opposed to this, but I’ll need some major help from others to implement it.

**Comment by mmahmoudian** [comment]
Perhaps the patch function can recursively fetch patches from github:
https://github.com/ismayc/thesisdown/commit/9bf78a3be301bcca0e9d7674e3259522dca268f7.patch
and then check **if they are applicable**, apply them.

**Comment by ismayc** [comment]
Yes, but what if someone, say, completely deleted parts of the YAML of `index.Rmd` for their thesis. Then when they apply the patch it will potentially replace things that they have changed even though they probably don't want that changed. So the patch would need to go line by line over each change made in the {thesisdown} template and ask the user to confirm the change. To further complicate things, what if users added white space or moved elements around. In that case, the patch would also need to go search for entries and then ask users to change those as well. This seems logistically really hard. Am I missing something?

**Comment by mmahmoudian** [comment]
Regarding the part that you said user have deleted parts of something, from my original post:
>  the function always checks "from what" and if it matched it will patch it. 

so if a line in the previous release was something and now should be updated, the code looks for the untouched line, if it finds the untouched line it will patch it, otherwise, perhaps message user just moves on.
About whitespace, the function can/should be agnostic about that. The package [stringi](https://cran.r-project.org/web/packages/stringi/index.html) can be useful for matching and patching.


Logically, like genome, there are parts of the package template that are conserved and is highly unlikely that a user changes them (probably those that are related to the functionality of this package and coupling with knitr and etc. Also, there are some parts that are very likely that user modify them like the first part of `index.Rmd`. But regardless of these, the patch always can check "what was in previous release" and replace it "what it is in the newer release".


To be clear, this is just a suggestion out of helplessness, as every time the package gets updated, I/user should create a new `index/` folder, diff their thesis folder with this newly created `index/´ and manually go through the changed lines most of which might be too LaTeX specific that the user does not necessarily know what they mean and how that can affect their work. In other words, If I was super comfortable with LaTeX I would have never picked thesisdown or other alternatives to write my thesis. I personally wrote my Masters thesis in LaTeX and the final finetuning the position of tables and figures was so painful that I decided to never do a massive writing project in it ever again.

**Comment by mmahmoudian** [comment]
Another alternative would be to have a separate config file and have the core parts of the template that are crucial for functionality separate. Then users can update and tinker with the config files and be more cautious about tinkering with the core parts. This way patching would be much easier as there is less interaction and touchings on essential parts. Also te patch function would know which files and parts it should not ever touch.

**Comment by ismayc** [comment]
Agreed on all points. I like the idea of a config file to further simplify things. Is this possible with YAML from `index.Rmd` and {bookdown}? It would be great to allow the Rmd files to be what the user can tweak as they wish and for the changes to happen elsewhere. That was what I wanted to do when {thesisdown} was originally created, but I couldn’t figure out a way to make that work. I’m open to PRs (as always) if there are more efficient ways to do this than what currently is in the package. I’d love to make the process simpler for anyone writing a thesis and to allow them to focus on writing as much as possible without having to worry about LaTeX and tweaking settings throughout.

**Comment by abichat** [comment]
As a user of {thesisdown} for my PhD manuscript and maintainer of a template, I thing the best option (compared to the amount of work the other option require) is to watch the original repository and to copy changes to our personal code. 
Three of the files you cited (bookdown.yml, index.Rmd and template.tex) are completely different from the original template. 
Moreover, changes usually happen because issue are fixed and 
* this does not happen really often (and I hope less and less) 
* the user could have already tweak his code to fix the problem 
Imho, he best thing to do is to write informative commit messages to help decide the final user if he should change his code, and properly reference issues & commits 
These are my personal reflexion about this issues, hope it helps in your thoughts

**Comment by ismayc** [comment]
Thanks for chiming in. One way to alleviate this is for me to add details into a NEWS.md file laying out all changes made to the package and which specific commits those correspond to. That’s best practice and should be a part of the workflow going forward as the package matures.

**Comment by abichat** [comment]
The `NEWS.md` is a good idea. If you want to go further, you can also add a dedicated file in `inst/` which explains the changes and what they fix (or it could be a very verbose `NEWS.md`).

**Comment by ismayc** [comment]
I’ve started by updating NEWS.md to v0.1.0. Could you review if the depth covered there will suffice or should I expand further? I can add this NEWS.md file into the template when downloaded too.

**Comment by mmahmoudian** [comment]
Thanks for the efforts, but I didn't find the `NEWS.md` useful at all!

From my point of view, the user, I would like to know which changes concern me. Meaning which ones I should merge into my thesis. This `NEWS.md` actually has some good changes listed:
1. changes to the R package (the hex sticker, the bookdown customizations, etc)
2. some changes to the text of the template files (https in the .bib  and .Rmd files,  etc.) and `README.md`, or [removing whitespaces](https://github.com/ismayc/thesisdown/commit/2a39610b8f3a97974d9220976edb1d14cd74d6e5#diff-a55cef02eeaf109b51e2070b11b60d34L65)
3. some changes to affect they layout ([such as R script text width](https://github.com/ismayc/thesisdown/commit/2a39610b8f3a97974d9220976edb1d14cd74d6e5#diff-a55cef02eeaf109b51e2070b11b60d34R93))

Out of these three, the third one is crucial for someone like me!! The 1 and 2 are changes that I am not (and should not be) interested in and they will never get merged into my in-progress thesis files.

Therefore, a better format of writing the `NEWS.md` imho is to put the changes in categories. Those that are essential to the user (updating the R package, applying the style changes) and those that are not essential but nevertheless are changes in this release.

-------

The way I merged it this time was to update the package, create a new thesis from the template in a temp folder, diff both folders line by line and file by file and manually go through changes in files other than my chapters. This was painful, but also I almost missed the relocation of `00--perlim.Rmd` and `00--abstract.Rmd` ! I'm not even sure this change has happened in this release or the previous one!

I know that you say this package rarely changes, but so far during the past week many things changed, for example the removal of chapter orders from the `_bookdown.yml` which I don't know how the orders are now defined in this version (it cannot be the leading numbers in the file names since there are ties like  `00--perlim.Rmd` and `00--abstract.Rmd`!

Ultimately, I think there are still room for improvement.

**Comment by ismayc** [comment]
Thanks for the feedback. The changes happened to fix warnings that you were seeing and reported as an issue last week :). In fixing that issue, some modifications were made to potentially fix other possible problems. If you look back at the commit history before this last week, you’ll see not much has changed over the last 4 years.

I’ll keep working on it and adding in sections to NEWS.md makes sense too. Of course, if you want to make suggestions or changes, you can also do so directly via a pull request to the package. That can help you to see the design phase head on and make your case to why others might be interested in changes you are suggesting.

**Comment by mmahmoudian** [comment]
Ok, I think I came up with a good suggestion. Perhaps this repo/R package should start going through the release-based system and the NEWS gets updated per release.

I checked the `remotes::install_github()` and it seems there is easy way to install [releases/latest](https://github.com/ismayc/thesisdown/releases/latest) through:

```R
remotes::install_github("ismayc/thesisdown@*release")
```

![image](https://user-images.githubusercontent.com/390889/83278293-09f2b700-a1dc-11ea-8efc-8b7953f3373c.png)

Therefore, we can instruct people to always update to the latest release which is always tested and has a complete NEWS. This way no one will accidentally downloads master branch which might contain some unfinished parts or it might not fully work.

Then the for each version in NEWS we can have 3 parts:
1. **R package**: Which user don't need to take any immediate action as it is something internal to the package (dependency upgrade, fine-tuning, etc.)
2. **Template Critical**: Changes to skeleton or .cls file that should be merged by user
3. **Template Non-critical**: Changes to Skeleton or README which user can skip merging

**Comment by ismayc** [comment]
Sounds good! This will become easier when the package gets on CRAN too. So the releases will match with that. Then `develop` will be used for making changes and testing and `master` will reflect the tested and working version of the package. Unit tests and ideally end-to-end tests will also need to be written, which is trickier with Rmds like this, but as the package develops it will help with all this to get the package into a more traditional workflow like other R packages.

I appreciate your patience with this and I know it can be frustrating to go back and make changes. Know that I’d like to make it better and am working on doing so.

**Comment by ismayc** [comment]
I'll plan to follow the release steps outlined in the [R Packages](http://r-pkgs.had.co.nz/release.html) book after 0.1.0 goes to CRAN in much the same way as other packages I've worked on ([infer](https://github.com/tidymodels/infer), [moderndive](https://github.com/moderndive/moderndive), and [fivethirtyeight](https://github.com/rudeboybert/fivethirtyeight)) have followed a similar release schedule type of flow. I'd like to see if anyone else detects any issues in the current state on `master` over the next month or so. I'm also hoping to implement a few new functions as well to the package before going to CRAN ideally along with unit tests. The NEWS.md file can also be viewed at http://thesisdown.rbind.io/news/index.html with the sections laid out now.

**Comment by ismayc** [comment]
I'm closing this issue for now, but feel free to reopen as needed.
